libraries/Microsoft.Bot.Configuration/BotConfigurationExtensions.cs

```diff
@@ -0,0 +1,153 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+namespace Microsoft.Bot.Configuration
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+
+    public static class BotConfigurationExtensions
+    {
+        /// <summary>
+        /// Returns a IEnumerable of a specific <see cref="ConnectedService"/> type contained in the bot configuration file.
+        /// </summary>
+        /// <param name="botServices">BotConfiguration's services.</param>
+        /// <param name="serviceType">The type of <see cref="ConnectedService"/> to search.</param>
+        /// <returns>List of bot's <see cref="ConnectedService"/>.</returns>
+        public static IEnumerable<ConnectedService> OfType(this List<ConnectedService> botServices, string serviceType)
+        {
+            // If botServices is null, throws an exception
+            if (botServices == null)
+            {
+                throw new ArgumentNullException();
+            }
+
+            // If there aren't any service, returns an empty IEnumerable
+            if (botServices.Count == 0)
+            {
+                return Enumerable.Empty<ConnectedService>();
+            }
+
+            // Obtains all the services of the specified type
+            var botFilteredServices = botServices.Where(service => service.Type.Equals(serviceType, StringComparison.CurrentCultureIgnoreCase));
+
+            // If there isn't any service of the specified type, returns an empty IEnumerable
+            if (botFilteredServices == null || botFilteredServices.Count() == 0)
```

Comment:
1. `Where` will never return `null`. All LINQ methods return a valid `IEnumerable<>`.
 2. `.Count()` is an `O(N)` operation. If anything you want to just be doing `!botFilteredServices.Any()`, but even this is unnecessary.

Again, this is short circuiting logic that is unnecessary and could be removed entirely and just return the result of the `Where` directly. It will always produce a valid set whether that set is empty or not.

My proposed final form of this method, if you keep it at all (per other comments), would be:

```CSharp
public static IEnumerable<ConnectedService> OfType(this IEnumerable<ConnectedService> botServices, string serviceType)
{
   // If botServices is null, throws an exception 




            if (botServices == null) 
            { 
                throw new ArgumentNullException(nameof(botServices)); 
            }

            return botServices.Where(service => service.Type.Equals(serviceType, StringComparison.OrdinalIgnoreCase));
}
````

Reply:
Finally, we removed this method.